src/boat_simulator/tests/unit/nodes/low_level_control/test_controller.py

```diff
@@ -0,0 +1,212 @@
+from boat_simulator.nodes.low_level_control.controller import (
+    RudderController,
+    SailController,
+)
+
+
+@pytest.fixture(
+    params=[
+        (60.3, 45, 0, 0.5, 0.7, 0.34, 2),
+        (-70.2, 30, 25, 0.5, 0.7, 0.34, 0.5),
+        (30.8, 20.4, 25, 1, 0.7, 0.34, 2),
+        (-24, -24, 9.2, 0.5, 0.7, 0.34, 2),
+        (90, -90, 30, 2, 0.7, 0.34, 1.5),
+        (50.488, -36.78, 30, 0.5, 0.7, 0.34, 1),
+        (40, -39, 10, 0.5, 0.7, 0.34, 0.5),
+        (0, 0, 0, 0.5, 0.7, 0.34, 1),
+    ]
+)
+def rudder_controller(request):
+    # Initialize the RudderController object with parameters from the fixture
+    current_heading, desired_heading, current_control_ang, time_step, kp, cp, control_speed = (
+        request.param
+    )
+    return RudderController(
+        current_heading=current_heading,
+        desired_heading=desired_heading,
+        current_control_ang=current_control_ang,
+        time_step=time_step,
+        kp=kp,
+        cp=cp,
+        control_speed=control_speed,
+    )
+
+
+def test_compute_error(rudder_controller):
+    # Test compute_error method of RudderController
+    error = rudder_controller._compute_error()
+    desired_rad = np.deg2rad(rudder_controller.desired_heading)
+    current_rad = np.deg2rad(rudder_controller.current_heading)
+    exp_error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+    assert np.equal(error, exp_error)
+
+
+def test_compute_setpoint(rudder_controller):
+    # Test compute_setpoint method of RudderController
+    setpoint = rudder_controller._compute_setpoint()
+    heading_error = rudder_controller._compute_error()  # heading_error in radians
+
+    rudder_setpoint = (rudder_controller.kp * heading_error) / (
+        1 + (rudder_controller.cp * abs(heading_error))
+    )
+
+    rudder_setpoint = min(
+        max(rudder_setpoint, np.deg2rad(rudder_controller.max_angle_range[0])),
+        np.deg2rad(rudder_controller.max_angle_range[1]),
+    )
+
+    rudder_setpoint_deg = np.rad2deg(rudder_setpoint)  # in degrees
+    running_error = rudder_setpoint_deg - rudder_controller.current_control_ang  # in degrees
+    assert np.equal(setpoint, rudder_setpoint_deg)
+    assert np.equal(running_error, rudder_controller.running_error)
+
+
+def test_reset_setpoint(rudder_controller):
+    # Test reset_setpoint method of RudderController
+    new_desired_heading = 60
+    new_current_heading = 30
+    expected = rudder_controller.reset_setpoint(new_desired_heading, new_current_heading)
+    rudder_controller._change_desired_heading(new_desired_heading)
+    calculated = rudder_controller._compute_setpoint()
+    np.equal(calculated, expected)
+
+
+# testing for one iteration only
+def test_update_state(rudder_controller):
+    current_control = rudder_controller.current_control_ang
+    rudder_controller._compute_setpoint()
+    assert rudder_controller.update_state() == (abs(rudder_controller.running_error) == 0)
```

Comment:
Wouldn't `assert rudder_controller.update_state()` be good enough here? I'm also a little confused on why running error needs to be zero here. Couldn't it be non-zero since we now call reset upon initialization? A single iteration shouldn't bring the running error to zero in general, right?

Reply:
Basically, update.state() will return false if we aren't at the target yet, so assert rudder_controller.update.state() will usually be false since running_error != 0. My check there is basically checking if update_state is false, that means the magnitude of running_error is greater than 0. If update_state is true, then running_error = 0. That was my way of checking the two possible cases of a single iteration, either moving towards our target angle or reaching our target angle.

Reply:
Oh I see, thanks for explaining